README
lib::byversion

So you have >1 Perl Installs. You have >1 Perl installs right? And you
switch between running them how?

Let me guess, somewhere you have code that sets a different value for
PERL5LIB depending on what Perl you're using. Oh you use "perlbrew"?
<http://grep.cpan.me/?q=PERL5LIB+dist=App-perlbrew>

This is a slightly different approach:

1. Set up your user-land PERL5LIB directories in a regular pattern
differing only by "perl" version
        $HOME/Foo/Bar/5.16.0/lib/...
        $HOME/Foo/Bar/5.16.1/lib/...
        $HOME/Foo/Bar/5.16.2/lib/...

2. Set the following in your %ENV
        PERL5OPT="-Mlib::byversion='$HOME/Foo/Bar/%V/lib/...'"

3. Done!
    The right PERL5LIB gets loaded based on which "perl" you use.

Yes, yes, catch 22, "lib::byversion" and its dependencies need to be in
your lib to start with.

O.k. That is a problem, slightly. But assuming you can get that in each
"perl" install somehow, you can load each "perl"'s user library
directories magically with this module once its loaded.

And "assuming you can get that in each "perl" install somehow" =~ with a
bit of luck, this feature or something like it might just be added to
Perl itself, as this is just a prototype idea to prove it works ( or as
the case may be, not ).

And even if that never happens, and you like this module, you can still
install this module into all your "perl"'s and keep a separate
"user-PERL5LIB-per-perl" without having to use lots of scripts to hold
it together, and for System Perls, you may even be fortunate enough to
get this module shipped by your "OS" of choice. Wouldn't that be dandy.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for lib-byversion

0.002003 2017-03-07T09:44:31Z 18f65f5

0.002002 2017-03-07T09:43:05Z 97a6c72
 [Dependencies::Stats]
 - Dependencies changed since 0.002001, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: ↓1

0.002001 2015-08-18T04:35:32Z fb6ecb3
 [Metadata]
 - CPAN.com? What planet is this?
 - Author fixed to be @cpan.org

0.002000 2015-08-18T03:45:00Z dc0f081
 [01 - Installer Changed]
 - Installer now migrated to use ExtUtils::MakeMaker
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +24 ↑2 -14 (recommends: +1, suggests: +2 -1)
 - test: +3 ↓1 (recommends: +3)

 [Tests]
 - More test coverage

0.001001 2013-11-20T06:54:37Z
 [00 Bugfix]
 - Fix test failing on <5.10

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +45 -1 (recommends: -1, suggests: +1 -1)
 - runtime: +1
 - test: ↑1 -2

 [Packaging]
 - use Module::Build test_requires
 - use new fallback code

 [Tests]
 - use Test::Compile::PerFile instead of Test::Compile
 - Regen ReportVersions::Tiny to hide devel prereqs
 - Regen Test::CPAN::Changes
 - Regen Test::Kwaltiee

0.001000 2013-02-15T21:10:15Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "add paths to \@INC depending on which version of Perl is running.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "lib-byversion",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "lib::byversion",
  "PREREQ_PM" => {
    "String::Formatter" => 0,
    "lib" => 0,
    "strict" => 0,
    "version" => "0.77",
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.002003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "String::Formatter" => 0,
  "Test::More" => 0,
  "lib" => 0,
  "strict" => 0,
  "version" => "0.77",
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/lib/byversion.pm
use 5.006;
use strict;
use warnings;

package lib::byversion;

our $VERSION = '0.002003';

# ABSTRACT: add paths to @INC depending on which version of Perl is running.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use lib ();
use version 0.77;

use String::Formatter stringf => {
  -as   => path_format =>,
  codes => {
    v => "$]",
    V => do {
      my $x = version->parse("$]")->normal;
      $x =~ s{^v}{}sx;
      $x;
    },
  },
};

sub import {
  my ( undef, @args ) = @_;
  if ( @args != 1 ) {
    die 'lib::byversion->import takes exactly one argument, instead, you specified ' . scalar @args;
  }
  my $path = path_format(@args);
  return lib->import($path);
}

## no critic (ProhibitBuiltinHomonyms)
sub unimport {
  my ( undef, @args ) = @_;
  if ( @args != 1 ) {
    die 'lib::byversion->unimport takes exactly one argument, instead, you specified ' . scalar @args;
  }
  my $path = path_format(@args);
  return lib->unimport($path);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

lib::byversion - add paths to @INC depending on which version of Perl is running.

=head1 VERSION

version 0.002003

=head1 SYNOPSIS

    PERL5OPT="-Mlib::byversion='$HOME/Foo/Bar/%V/lib/...'"

or alternatively

    use lib::byversion "/some/path/%V/lib/...";

=head1 DESCRIPTION

So you have >1 Perl Installs.  You have >1 Perl installs right?
And you switch between running them how?

Let me guess, somewhere you have code that sets a different value for PERL5LIB
depending on what Perl you're using.
Oh you use L<< C<perlbrew>?|http://grep.cpan.me/?q=PERL5LIB+dist=App-perlbrew >>

This is a slightly different approach:

=over 4

=item 1. Set up your user-land PERL5LIB directories in a regular pattern
differing only by C<perl> version

    $HOME/Foo/Bar/5.16.0/lib/...
    $HOME/Foo/Bar/5.16.1/lib/...
    $HOME/Foo/Bar/5.16.2/lib/...

=item 2. Set the following in your C<%ENV>

    PERL5OPT="-Mlib::byversion='$HOME/Foo/Bar/%V/lib/...'"

=item 3. Done!

The right PERL5LIB gets loaded based on which C<perl> you use.

=back

Yes, yes, catch 22, C<lib::byversion> and its dependencies need to be in
your lib to start with.

O.k. That is a problem, slightly. But assuming you can get that in each
C<perl> install somehow, you can load each C<perl>'s user library directories
magically with this module once its loaded.

And "assuming you can get that in each C<perl> install somehow" =~ with a bit
of luck, this feature or something like it might just be added to Perl itself,
as this is just a prototype idea to prove it works ( or as the case may be,
not ).

And even if that never happens, and you like this module, you can still
install this module into all your C<perl>'s and keep a separate
C<user-PERL5LIB-per-perl> without having to use lots of scripts to hold it
together, and for System Perls, you may even be fortunate enough to get this
module shipped by your C<OS> of choice. Wouldn't that be dandy.

=head1 IMPORT

    use lib::byversion $param
    lib::byversion->import($param)
    perl -Mlib::byversion=$param

etc.

C<lib::byversion> expects one parameter, a string path containing templated
variables for versions.

Current defined parameters include:

=over 4

=item C<%V>

This is an analogue of C<$^V> except :

=over 4

=item it should work on even C<perl>s that didn't have C<$^V>, as it converts
it from C<$]> with L<version.pm|version>

=item it lacks the preceding C<v>, because this is more usually what you want
and its easier to template it in than take it out.

=back

Example:

    %V = 5.16.9

=item C<%v>

This is the same as C<$]> on your Perl.

Example:

    %v = 5.016009

=back

More may be slated at some future time, e.g.: to allow support for components
based on C<git> C<sha1>'s, but I figured to upload something that works before
I bloat it out with features nobody will ever use.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    templated
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

$bundle->add_or_append_policy_field( 'Subroutines::ProhibitCallsToUndeclaredSubs' =>
    ( 'exempt_subs' => [ 'String::Formatter::str_rf', 'String::Formatter::path_format', ] ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('Subroutines::RequireFinalReturn');
#$bundle->remove_policy('Subroutines::RequireArgUnpacking');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.039.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4214"
  },
  "configure_requires" => {
    "Module::Build" => "0.4214",
    "perl" => "5.006"
  },
  "dist_abstract" => "add paths to \@INC depending on which version of C<perl> is running.",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "lib-byversion",
  "dist_version" => "0.001002",
  "license" => "perl",
  "module_name" => "lib::byversion",
  "recursive_test_files" => 1,
  "requires" => {
    "String::Formatter" => 0,
    "lib" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "version" => "0.77",
    "warnings" => 0
  },
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.302007_004",
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4214",
  "Test::More" => "1.302007_004",
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
